Feat: Custom matchers

brooksbecton opened the pull request:
### Summary

Begins work on custom jest matchers as part of issue #99 

### Test plan

`yarn test`

Esemesek commented:
@thymikee I don't think we want this `package-lock.json` committed.

thymikee commented on src/__tests__/custom-matchers.test.js:
```diff
@@ -0,0 +1,65 @@
+import React from 'react';
```
Please add `// @flow` pragma

thymikee commented on src/custom-matchers/extendExpect.js:
```diff
@@ -0,0 +1,3 @@
+import extensions from './index';
```
`// @flow`

thymikee commented on src/custom-matchers/index.js:
```diff
@@ -0,0 +1,4 @@
+import { toBeDisabled } from './toBeDisabled';
```
`// @flow`

thymikee commented on src/__tests__/custom-matchers.test.js:
```diff
@@ -0,0 +1,65 @@
+import { create } from 'react-test-renderer';
+
+describe('custom-matchers', () => {
+  function render(node) {
```
can we use `render` from this library?

brooksbecton commented:
- Removed the package-lock :sweat_smile: 
- Added `// @flow`
- Using `react-native-testing-library`'s render function 

I found Esemesek's branch with the custom-matchers here: https://github.com/callstack/react-native-testing-library/tree/Esemesek/custom-matchers/src/matchers

and I used most of it's .toBeEmpty logic because it is way more thorough

brooksbecton commented:
Removed package-lock.json. 

Were ya'll wanting to get most of the matchers that are in jest-dom in this PR?

thymikee commented:
I'd rather like to start small :)

thymikee commented on src/__tests__/custom-matchers.test.js:
```diff
@@ -0,0 +1,125 @@
+        <View>
+          <Button
+            testID={BUTTON_EXPLICIT_DISABLED}
+            disabled={true}
```
This case can be removed, it's the same as below from jsx point of view

brooksbecton replied:
Removed the 'BUTTON_EXPLICIT_DISABLED' test case

thymikee commented on src/custom-matchers/toBeEmpty.js:
```diff
@@ -0,0 +1,28 @@
+// @flow
+
+function isEmpty(instance: ReactTestInstance) {
+  if (instance.props.children instanceof Array) {
```
We should use React.Children for iteration over children

brooksbecton replied:
Are you saying we should check to see if it is instanceof React.Children? 
I don't understand

thymikee commented:
Thank you so much for your time! I'll close this in favor of #298 though. We're going to use `@testing-library/jest-native`. I hope it will only be a matter of documenting it and providing some tests that we're compatible :). If we want to add more matchers, we can contribute it back there :)
